cql3/statements/modification_statement.cc

```diff
@@ -272,24 +270,44 @@ modification_statement::do_execute(query_processor& qp, service::query_state& qs
     _restrictions->validate_primary_key(options);
 
     if (has_conditions()) {
-        return execute_with_condition(qp, qs, options);
+        co_return co_await execute_with_condition(qp, qs, options);
     }
 
-    return execute_without_condition(qp, qs, options).then([] (coordinator_result<> res) {
-        if (!res) {
-            return make_ready_future<::shared_ptr<cql_transport::messages::result_message>>(
-                    seastar::make_shared<cql_transport::messages::result_message::exception>(std::move(res).assume_error()));
+    json_cache_opt json_cache = maybe_prepare_json_cache(options);
+    std::vector<dht::partition_range> keys = build_partition_keys(options, json_cache);
+
+    auto res = co_await execute_without_condition(qp, qs, options, json_cache, keys);
+    
+    if (!res) {
+        co_return co_await make_ready_future<::shared_ptr<cql_transport::messages::result_message>>(
+                seastar::make_shared<cql_transport::messages::result_message::exception>(std::move(res).assume_error()));
+    }
+
+    if (keys.size() != 1) {
```

Comment:
execute_without_condition() takes keys by non-const & and get_mutations() moves it, so here keys will be always empty. We need to move this logic earlier.

Reply:
We really need unit tests here.

Reply:
Where should tests be placed? And what exactly should they check?

Reply:
https://github.com/scylladb/scylladb/pull/15410#issuecomment-1724234812

We should check the matrix {token-aware, not-token-aware} X {routed correctly, routed-incorrectly} X {select, update}, and verify appropriately that routing info is {present, missing}.

e.g.  For any not-token-aware request the info should be missing, for any token-aware routed correctly the info should me missing, etc.